socketcan/libcanard/src/socketcan.c

```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
     const int16_t poll_result = doPoll(fd, POLLIN, timeout_usec);
     if (poll_result > 0)
     {
-        struct timespec ts;
-        if (clock_gettime(CLOCK_TAI, &ts) < 0)  // TODO: request the timestamp from the kernel.
+        // Initialize the message header scatter/gather array.
+        // We use the CAN FD struct regardless of whether the CAN FD socket option is set.
+        // Per the user manual, this is acceptable because they are binary compatible.
+        struct iovec       iov           = {0};                    // Scatter/gather array items struct.
+        struct canfd_frame sockcan_frame = {0};                    // CAN FD frame storage.
+        iov.iov_base                     = &sockcan_frame;         // Starting address.
+        iov.iov_len                      = sizeof(sockcan_frame);  // Number of bytes to transfer.
+
+        // Determine the size of the ancillary data and zero-initialize the buffer for it.
+        // We require space for both the receive message and the time stamp.
+        // TODO: Aligned storage is used. --> Don't know how to do this in a portable way....
+        // TODO: C complains that the control_size is not known at compile time, hence the use of malloc/free.
+        //       Should not be needed, as those structs are fixed-size?? Ah, cmsghdr has a flexarray extension in c99?
```

Comment:
Yep. Why don't you just follow the recommendation given in the docs here: https://man7.org/linux/man-pages/man3/cmsg.3.html#EXAMPLES (tl;dr: use union).

Reply:
Learned something new. Also, turns-out older manpages don't have this detailed example.